cpp/memilio/io/epi_data.h

```diff
@@ -0,0 +1,460 @@
+ * state_id and county_id will be empty.
+ */
+class DiviEntry 
+{
+public:
+    double num_icu;
+    Date date;
+    boost::optional<regions::de::StateId> state_id;
+    boost::optional<regions::de::CountyId> county_id;    
+
+    template <class IoContext>
+    static IOResult<DiviEntry> deserialize(IoContext& io)
+    {
+        auto obj       = io.expect_object("RkiEntry");
+        auto num_icu   = obj.expect_element("ICU", Tag<double>{});
+        auto date      = obj.expect_element("Date", Tag<StringDate>{});
+        auto state_id  = obj.expect_optional("ID_State", Tag<regions::de::StateId>{});
+        auto county_id = obj.expect_optional("ID_County", Tag<regions::de::CountyId>{});
+        return apply(
+            io,
+            [](auto&& ni, auto&& d, auto&& sid, auto&& cid) {
+                return DiviEntry{ni, d, sid, cid};
+            },
+            num_icu, date, state_id, county_id);
+    }
+};
+
+/**
+ * Deserialize a list of DiviEntry from json.
+ * @param jsvalue Json value that contains DIVI data.
+ * @return list of DiviEntry.
+ */
+inline IOResult<std::vector<DiviEntry>> deserialize_divi_data(const Json::Value& jsvalue) {
+    return deserialize_json(jsvalue, Tag<std::vector<DiviEntry>>{});
+}
+
+/**
+ * deserialize a list of DiviEntry from json.
+ * @param filename Json file that contains DIVI data.
+ * @return list of DiviEntry.
+ */
+inline IOResult<std::vector<DiviEntry>> read_divi_data(const std::string& filename)
+{
+    return read_json(filename, Tag<std::vector<DiviEntry>>{});
+}
+
+namespace details {
+    //check all results in a vector and unpack each
+    template<class T>
+    IOResult<std::vector<T>> unpack_all(const std::vector<IOResult<T>>& v) {
+        std::vector<T> w;
+        w.reserve(v.size());
+        for (auto&& r : v) {
+            BOOST_OUTCOME_TRY(t, r);
+            w.push_back(t);
+        }
+        return success(w);
+    }
+}
+
+/**
+ * Represents an entry of a population data file.
+ * Population per age group in a region.
+ * Region can be a county, a state, or a country. If it is a country, both
+ * state_id and county_id will be empty.
+ */
+class PopulationDataEntry
+{
+public:
+    static const std::array<const char*, 11> age_group_names;
+
+    CustomIndexArray<double, AgeGroup> population;
+    boost::optional<regions::de::StateId> state_id;
+    boost::optional<regions::de::CountyId> county_id;    
+    
+    template <class IoContext>
+    static IOResult<PopulationDataEntry> deserialize(IoContext& io)
+    {
+        auto obj       = io.expect_object("RkiEntry");
+        auto state_id  = obj.expect_optional("ID_State", Tag<regions::de::StateId>{});
+        auto county_id = obj.expect_optional("ID_County", Tag<regions::de::CountyId>{});
+        std::vector<IOResult<double>> age_groups;
+        age_groups.reserve(age_group_names.size());
+        std::transform(age_group_names.begin(), age_group_names.end(), std::back_inserter(age_groups), [&obj](auto&& age_name) {
```

Comment:
Yes, it needs to be. For Json, it doesn't strictly need to be a reference since everything is copyable safely while using the same underlying storage for errors and file data. But this is not true in general, e.g. other file formats.